src/BoundsInference.cpp

```diff
@@ -109,27 +110,127 @@ class BoundsInference : public IRMutator {
+                << "Init definition shouldn't have a RDom\n";
+
+            vector<CondValue> result;
+
+            // Default case (no specialization)
+            vector<Expr> predicates;
+            if (def.domain().defined()) {
+                rdoms.insert(def.domain());
+                predicates = def.domain().split_predicate();
+            }
+            for (const Expr &val : def.values()) {
+                if (!predicates.empty()) {
+                    Expr cond_val = Call::make(val.type(),
+                                               Internal::Call::if_then_else,
+                                               {likely(predicates[0]), val, make_zero(val.type())},
+                                               Internal::Call::PureIntrinsic);
+                    for (size_t i = 1; i < predicates.size(); ++i) {
+                        cond_val = Call::make(cond_val.type(),
+                                              Internal::Call::if_then_else,
+                                              {likely(predicates[i]), cond_val, make_zero(cond_val.type())},
+                                              Internal::Call::PureIntrinsic);
+                    }
+                    result.push_back(CondValue(const_true(), cond_val));
+                } else {
+                    result.push_back(CondValue(const_true(), val));
+                }
+            }
+            if (is_update) {
+                for (const Expr &arg : def.args()) {
+                    if (!predicates.empty()) {
+                        Expr cond_arg = Call::make(arg.type(),
+                                                   Internal::Call::if_then_else,
+                                                   {likely(predicates[0]), arg, make_zero(arg.type())},
+                                                   Internal::Call::PureIntrinsic);
+                        for (size_t i = 1; i < predicates.size(); ++i) {
+                            cond_arg = Call::make(cond_arg.type(),
+                                                  Internal::Call::if_then_else,
+                                                  {likely(predicates[i]), cond_arg, make_zero(cond_arg.type())},
+                                                  Internal::Call::PureIntrinsic);
+                        }
+                        result.push_back(CondValue(const_true(), cond_arg));
+                    } else {
+                        result.push_back(CondValue(const_true(), arg));
+                    }
+                }
+            }
+
+            const vector<Specialization> &specializations = def.specializations();
+            for (size_t i = specializations.size(); i > 0; i--) {
+                Expr s_cond = specializations[i-1].condition;
+                const Definition &s_def = specializations[i-1].definition;
+
+                // Else case (i.e. specialization condition is false)
+                for (CondValue &i : result) {
```

Comment:
I don't understand this code. You mutate i.cond by &&ing it with all the !s_conds, and then later you && it with all the s_conds.... isn't the result false?